### PR TITLE
Implement TopicMetadata for MessageRouter in C++ client

### PR DESCRIPTION
**Pull request by Licht-T**
### Motivation

Same motivation as https://github.com/apache/incubator-pulsar/pull/1025.

### Modifications

- Implement `TopicMetadata` in C++ client
- Add tests for Messege Router in C++ client

### Result

- We can pass additional information when `getPartition`.

### CODE REVIEW

File: pulsar-client-cpp/include/pulsar/Message.h [hunk 1]
```diff
@@ -44,14 +44,15 @@ class Message {
...
+    virtual ~Message() {}
 
     /**
      * Return the properties attached to the message.
      * Properties are application defined key/value pairs that will be attached to the message
      *
      * @return an unmodifiable view of the properties map
      */
-    const StringMap& getProperties() const;
+    virtual const StringMap& getProperties() const;
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Message.h]: Why is this change needed? We shouldn't have more types of messages at this point. 

Also, I think this is not binary compatible with current version and would have to be deferred to major version change.

File: pulsar-client-cpp/include/pulsar/MessageRoutingPolicy.h [hunk 2]
```diff
@@ -33,7 +34,7 @@ class MessageRoutingPolicy {
  public:
     virtual ~MessageRoutingPolicy() {}
 
-    virtual int getPartition(const Message& msg) = 0;
+    virtual int getPartition(const Message& msg, const TopicMetadata& topicMetadata) = 0;
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/MessageRoutingPolicy.h]: The original method needs to be there as since it's part of the user API. We could mark it as deprecated and remove it in 2.0. 

If the user was defining a custom router, it should continue working after this change

File: pulsar-client-cpp/lib/PartitionedProducerImpl.cc [hunk 3]
```diff
@@ -37,18 +39,28 @@ namespace pulsar {
                                                      const ProducerConfiguration& config):client_(client),
                                                                                           destinationName_(destinationName),
                                                                                           topic_(destinationName_->toString()),
-                                                                                          numPartitions_(numPartitions),
                                                                                           conf_(config),
                                                                                           state_(Pending)
     {
         numProducersCreated_ = 0;
+
+        topicMetadata_ = std::unique_ptr<TopicMetadata>(new TopicMetadataImpl(numPartitions));
```
Comment (merlimat) [inline pulsar-client-cpp/lib/PartitionedProducerImpl.cc]: In general we cannot yet use all the c++11 feature in general lib. We could use a `boost::scoped_ptr` here.

File: pulsar-client-cpp/tests/CMakeLists.txt [hunk 4]
```diff
@@ -17,12 +17,12 @@
 # under the License.
 #
 
-find_library(GTEST_LIBRARY_PATH gtest)
+find_library(GMOCK_LIBRARY_PATH gmock)
```
Comment (merlimat) [inline pulsar-client-cpp/tests/CMakeLists.txt]: We might also need to update the Docker images used for building the C++ libray as well as the documentation.

**Comment by Licht-T** [comment]
retest this please

File: pulsar-client-cpp/include/pulsar/MessageRoutingPolicy.h (hunk 2 above)
Comment (Licht-T) [reply pulsar-client-cpp/include/pulsar/MessageRoutingPolicy.h]: @merlimat We need to set these method to `virtual` because of the restriction of Google Mock.
https://github.com/google/googletest/blob/master/googlemock/docs/ForDummies.md#how-to-define-it

File: pulsar-client-cpp/include/pulsar/Message.h (hunk 1 above)
Comment (Licht-T) [reply pulsar-client-cpp/include/pulsar/Message.h]: @merlimat We need to set these method to virtual because of the restriction of Google Mock.
https://github.com/google/googletest/blob/master/googlemock/docs/ForDummies.md#how-to-define-it

**Comment by Licht-T** [comment]
@merlimat Almost fixed. What should I do on the mock test.

**Comment by Licht-T** [comment]
CI failed due to the old Docker image.

File: pulsar-client-cpp/include/pulsar/MessageRoutingPolicy.h [hunk 5]
```diff
@@ -33,7 +34,10 @@ class MessageRoutingPolicy {
  public:
     virtual ~MessageRoutingPolicy() {}
 
+    /** @deprecated */
     virtual int getPartition(const Message& msg) = 0;
+
+    virtual int getPartition(const Message& msg, const TopicMetadata& topicMetadata) = 0;
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/MessageRoutingPolicy.h]: I think this will break the compatibility for existing apps that have the custom router. 

For example, if I had : 
```cpp
class MyRouter : public MessageRoutingPolicy {
public:
  int getPartition(const Message& msg) { 
     // My custom routing logic
  }
};
```
This  will get a link error at runtime with the new Pulsar library.

I think the new method should have a default implementation that uses the other one, eg: 

```cpp
virtual int getPartition(const Message& msg, const TopicMetadata& topicMetadata)  {
  return getPartition(msg);
}
```

At the same time, I'm not sure we should leave the old `int getPartition(const Message& msg)` with no implementation, because that would force to provide both implementations to the user. 

Maybe we could default it to an invalid partition, like : 

```cpp
/** @deprecated 
   Use int getPartition(const Message& msg, const TopicMetadata& topicMetadata)
 */
virtual int getPartition(const Message& msg) {
   return -1;
}

**Comment by merlimat** [comment]
> @merlimat Almost fixed. What should I do on the mock test? 

I think we cannot break the binary compatibility at this point. We could do that once we move to Pulsar 2.0 (which should happen soon), but in the meantime we shouldn't be changing the method signatures.

**Comment by Licht-T** [comment]
Thanks @merlimat!
- The deprecated virtual method `int getPartition(const Message&)` is now throws `DeprecatedException`.
- Reverted the change in Message class and mock tests is disabled.

**Comment by merlimat** [comment]
Updated the build image. 

retest this please

**Comment by Licht-T** [comment]
retest.this please

**Comment by Licht-T** [comment]
Thanks @merlimat, now all green!

**Comment by Licht-T** [comment]
@merlimat Would you please merge into master? I want to do another work after this merged.

**Comment by merlimat** [comment]
> @merlimat Would you please merge into master? I want to do another work after this merged.

Yup, I lost track of this PR...

**Comment by merlimat** [comment]
Thanks!
